File: 자료구조/Sort/sort.java
// 정렬을 부수자.

import java.util.ArrayDeque;
import java.util.Arrays;

public class sort {
    
    public static void bubbleSort(String[] args) {
        int []arr = {7, 5, 9, 0, 3, 1, 6, 2, 4, 8};
        for (int i=0; i<arr.length-1; i++) { // 각 pass마다 맨 오른쪽에 max가 오게 됨
            boolean swapped = false;
            for (int j=0; j<arr.length-i; j++) { // 각 pass마다 맨 오른쪽 볼 필요 없어짐
                if (arr[j] > arr[j+1]) {
                    int temp = arr[j];
                    arr[j] = arr[j+1];
                    arr[j+1] = temp;
                    swapped = true;
                }

            // 해당 pass에서 swap발생x시, 이미 정렬된 상태
            if (swapped == false) break;
            }
        }

    }

    public static void selectionSort(String[] args) {
        
        int []arr = {7, 5, 9, 0, 3, 1, 6, 2, 4, 8};
        
        for (int i=0; i<arr.length-1; i++) { // pivot
            int minIndex = i; // 가장 작은 원소 인덱스
            for (int j=i+1; j<arr.length; j++) { // min 위치 찾기
                if (arr[minIndex] > arr[j]) minIndex = j;
            }

            // pivot과 min swap
            int tmp = arr[i];
            arr[i] = arr[minIndex];
            arr[minIndex] = tmp;
        }
        System.out.println(Arrays.toString(arr));
    }

    public static void insertionSort(String[] args) {
        
        int []arr = {7, 5, 9, 0, 3, 1, 6, 2, 4, 8};
        
        
        for(int i=1; i<arr.length; i++){ // pivot
            for(int j=i; j>=1; j--){ // pivot 왼편, sortedList
                
                if(arr[j] < arr[j-1]){  // sortedList의 원소가 원래 pivot원소보다 크면, 한 칸씩 왼쪽으로 이동
                    int tmp = arr[j];
                    arr[j] = arr[j-1];
                    arr[j-1] = tmp;
                }else break;  //자기보다 작은 데이터를 만나면 그 위치에서 멈춤
                
            }
        }
        System.out.println(Arrays.toString(arr));
    }

    public static void shellSort(String[] args) {

        int []arr = {7, 5, 9, 0, 3, 1, 6, 2, 4, 8};
        int n = arr.length;

        for (int gap=n/2; gap>0; gap/=2) { // gap: n/2부터 1이 될 때까지 /2
            for (int i=gap; i<n; i++) { // pivot: gap부터 n이 될 때까지 +1
                int j;
                int pivot = arr[i];

                // pivot왼쪽 수가 pivot보다 작으면 끝. gap씩 빼며 pivot에 의해 밀려남
                for (j=i-gap; j>=0&&arr[j]>pivot; j-=gap) {
                    arr[j+gap] = arr[j];
                }
                arr[j+gap] = pivot;
            }
        }

        for (int a: arr) {
            System.out.println(a+" ");
        }

    }

    public static int[] mergeSort(int[] arr) {
        if (arr.length < 2) return arr;

        int mid = arr.length/2;
        int[] leftArr = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] rightArr = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));

        int[] result = new int[arr.length];
        int re=0, l=0, r=0; // 결과 배열의 idx, left배열의 idx, right배열의 idx
        while (l<leftArr.length && r<rightArr.length) {
            if (leftArr[l]< rightArr[r]) result[re++] = leftArr[l++];
            else result[re++] = rightArr[r++];
        } 
        while (l<leftArr.length) result[re++] = leftArr[l++];
        while (r<rightArr.length) result[re++] = rightArr[r++];
        return result;
    }

    // in-place update
    public static int[] mergeSortOptimized(int[] arr) {
        div(arr, 0, arr.length);
        return arr;
    }
    // 봐야할 idx만 전달 -> 범위 나누며 쪼갬
    private static void div(int[] arr, int low, int high) {
        if (high-low<2) return;
        int mid = (low+high)/2;
        div(arr, low, mid); // 왼쪽 arr: low~mid
        div(arr, mid, high); // 오른쪽 arr: mid~high
        merge(arr, low, mid, high);
    }
    private static void merge(int[] arr, int low, int mid, int high) {
        int[] temp = new int[high-low];
        int t=0, l=low, r=mid; // temp의 idx, left:low~mid, right:mid~high

        while (l<mid && r<high) {
            if (arr[l]<arr[r]) temp[t++] = arr[l++];
            else temp[t++] = arr[r++];
        }
        // 털기
        while (l<mid) temp[t++] = arr[l++];
        while (r<high) temp[t++] = arr[r++];
        // # x가 [3,4,5,2,3,4]일 때, low1~3high면, temp=[4,5]로 잘림
        // # x[low] = temp[low-low]
        // # x[high] = temp[high-low]
        for (int i=low; i<high; i++) {
            arr[i] = temp[i-low];
        }
    }

    public static void radixSort(String args[]) {

        int arr[] = {7, 5 , 9, 0, 3, 1, 6, 2, 9, 1, 4, 8, 0, 5, 2};
        
        for (int pos=1; pos<=10000; pos*=10) { // 1의 자릿수부터 시작
            arr = queSort(arr, pos);
        }
        for (int n:arr) {
            System.out.println(n);
        }
    }
    private static int[] queSort(int[] arr, int pos) {
        int[] out = new int[arr.length]; // 현재 자릿수에 따른 정렬 후 결과
        ArrayDeque[] bucket = new ArrayDeque[10]; // 특정 자릿수의 0-9 bucket
        for (int i=0; i<10; i++) {
            bucket[i] = new ArrayDeque<Integer>();
        }
        
        for (int i=0; i<arr.length; i++) {
            int digit = (arr[i]/pos) % 10; // pos자릿수의 숫자
            bucket[digit].add(arr[i]);
        }
        
        int i=0, idx=0;
        while (i<10) {
            if (!bucket[i].isEmpty()) out[idx++] = (int)bucket[i].poll();
            else i++; // 다음 큐 보자 0-9
        }
        return out;
    }

    public static void countSort(String args[]){ // 계수 정렬
        
        //모든 원소의 값이 0보다 크거나 같다고 가정
        int arr[] = {7, 5 , 9, 0, 3, 1, 6, 2, 9, 1, 4, 8, 0, 5, 2};
        
        //모든 범위를 포함하는 리스트 선언(모든 값은 0으로 초기화, int는 기본적으로 0으로 초기호)
        int []count = new int [Arrays.stream(arr).max().getAsInt()];
        int []result = new int[arr.length];

        for (int i=0; i<arr.length; i++) {
            count[arr[i]] += 1; //각 데이터에 해당하는 인덱스의 값 증가
        }


        for (int i=1; i<count.length; i++) {
            count[i] += count[i-1]; // 빈도수 누적합
        }

        for (int i=arr.length-1; i>=0; i--) {
            int val = arr[i];
            count[val]--;
            result[count[val]] = val;
        }


        for(int i = 0; i < result.length; i++) {
			System.out.print(result[i] + "\t");
		}
    }
}

File: 자료구조/Sort/sort.py
def bubbleSort(x):
    for i in range(len(x)-1): # 각 pass. 오른쪽부터 max채워짐
        swapped = False
        for j in range(len(x)-i): # 이전 pass에서 오른쪽 max가 확정됨. 범위 감소
            if x[j]>x[j+1]:
                x[j], x[j+1] = x[j+1], x[j]
                swapped = True
        if not swapped: # 해당 pass에서 swap 발생x
            break # 이미 정렬된 상태

def selectionSort(x): # 최솟값을 맨 왼쪽에 배치
    for i in range(len(x)-1):
        min_idx = i
        for j in range(i+1, len(x)): # 최솟값의 idx 갱신
            if x[j]<x[min_idx]:
                min_idx = j
        x[i], x[min_idx] = x[min_idx], x[i]
      
def insertionSort(x):
    for i in range(1, len(x)): # pivot이 왼쪽 sortedList보다 크면 정렬 끝
        j, pivot = i-1, x[i]
        while pivot < x[j]: # 원래 x[j]를 j+1위치로 밀어내고, pivot보다 작은 원소 나올 때까지 j-1
            x[j+1] = x[j]
            j = j-1
        x[j+1] = pivot
    return x

def shellSort(x): # gap: n/2->1될때까지 insertionSort
    n = len(x)
    gap = n//2
    while gap>0: # 1이 될 때까지
        for i in range(gap, n): # pivot
            pivot = x[i]
            j = i-gap # i와 gap만큼 떨어진 왼쪽 원소
            while j>=0 and x[j]>pivot: # pivot보다 왼쪽 원소가 크면 swap
                x[j+gap] = x[j] # 원래 x[j]를 j+gap위치로 밀어내고, pivot보다 작은 원소 나올 때까지 j-gap
                j -= gap
            x[j+gap] = pivot
        gap //= 2
    return x

def heapSort(x):
    def heapify(arr, parent, size): # subtree-> max heap으로
        largest = parent
        left = 2*parent+1
        right = 2*parent+2
        
        if left<size and arr[left]>arr[largest]:
            largest = left
        if right<size and arr[right]>arr[largest]:
            largest = right
            
        if largest != parent: # 부모보다 자식이 크면 swap
            arr[largest], arr[parent] = arr[parent], arr[largest]
            heapify(arr, largest, size) # 자식 idx level로 넘어감
            
    n = len(x)
    # 최대 heap 생성 -> 루트가 서브트리 노드보다 큼 보장
    for i in range(n//2-1, -1, -1): # 맨마지막 원소의 루트부터 전달(n은 전체 노드 수)
        heapify(x, i, n)
    
    # root와 마지막 노드 swap후, 마지막 노드(제일 큰 수)빼고 heap화
    for i in range(n-1, 0, -1):
        x[i], x[0] = x[0], x[i]
        heapify(x, 0, i)

def mergeSort(x):
    if len(x)<2:
        return x
    mid = len(x)//2 # 기준 idx
    left_list = mergeSort(x[:mid]) # 정렬된 부분 list
    right_list = mergeSort(x[mid:])
    
    result = []
    l = r = 0 # 정렬된 두 부분 list 비교를 위한 포인터
    while l<len(left_list) and r<len(right_list):
        if left_list[l] < right_list[r]:
            result.append(left_list[l])
            l += 1
        else:
            result.append(right_list[r])
            r += 1
    # 둘 중 하나의 리스트는 아직 원소 남아 있음
    result += left_list[l:]
    result += right_list[r:]
    return result

def mergeSortOptimized(x): # result 따로 만들지 말고 inplace업데이트
    def sort(l, r): # 비교해야 하는 idx만 전달
        if r-l < 2:
            return
        mid = (l+r)//2
        sort(l, mid)
        sort(mid, r)
        merge(l, mid, r) # 두 인접 리스트 합치자
        
    def merge(low, mid, high): # 두 인접 리스트: low~mid, mid~high로 나뉨
        temp = []
        l, r = low, mid # l:low~mid, r:mid~high
        while l<mid and r<high:
            if x[l] < x[r]:
                temp.append(x[l])
                l += 1
            else:
                temp.append(x[r])
                r += 1
                
        while l<mid: # 털기
            temp.append(x[l]) 
            l += 1
        while r<high:
            temp.append(x[r])
            r += 1
        print(low, mid, high, x, temp)
        # x가 [3,4,5,2,3,4]일 때, low1~3high면, temp=[4,5]로 잘림
        # x[low] = temp[low-low]
        # x[high] = temp[high-low]
        for i in range(low, high):
            x[i] = temp[i-low]
        
    sort(0, len(x))
    return x # in-place update
print(mergeSortOptimized([3,4,5,2,3,4]))

from collections import deque
def radixSort(x):
    bucket = [deque() for _ in range(10)] # 0-9까지 큐
    max_num = max(x)
    que = deque(x)
    pos = 1 # 1의 자리 수부터 보기 위함

    while max_num >= pos: # 가장 큰 수의 자릿수 뛰어 넘으면 stop
        while que:
            num = que.popleft() # que에 들어간 순서대로 pop FIFO
            bucket[(num//pos) % 10].append(num) # 현재 자릿수 숫자에 해당하는 큐에 넣음
        
        for b in bucket: # 0-9 큐
            while b:
                que.append(b.popleft()) # 해당 자릿수에서 작았던 숫자부터 QUE에 들어감
        print(pos, "자릿 수 정렬: ", list(que))        
        pos *= 10 # 자릿수 증가
    return list(que)
print(radixSort([123,2154,222,4,283]))

def countSort(x):
    count = dict()
    for num in x: # 해당 원소가 list에 몇 개 있는지
        if num in count:
            count[num] += 1
        else:
            count[num] = 1
    
    result = [0]*len(x)
    for num in x:
        count[num] -= 1
        result[count[num]] = num
    return result
